
| <h1> Activity 1.1 </h1> | <h1> Python Challenge </h1> |
|--- | --- |
Name: Christian Jay L. Cuevas
Course and Section: CPE019S3 <br>
Date Submitted: 1/24/2024 <br>
Instructor: Engr. Roman Richard
Date Performed: 1/24/2024
Date Submitted: 1/24/2024

<hr>

**Instructions**: Answer the questions or complete the tasks outlined below, use the specific method described if applicable. Insert a code cell after each text cell.

**Objectives:**
- Test understanding of Python basics
- Review Python basic to intermediate commands

<hr>

1. What is 3 to the power of 5?

In [ ]:
 3**5

243

###Observation:

I observed that the expression ** is used to assign an exponent to a number. We can observe that 3**5 is equivalent to 3^5 in mathematics and the answer to that is 243

2.  Create a variable, s, containing the string "This course is amazing!". Using the variable, split the string into a list.

In [ ]:
s = "This course is amazing!"
s.split()

['This', 'course', 'is', 'amazing!']

###Observation:

The string stored in the variable s is "This course is amazing!", when the function .split() was applied to the variable s, it became a list. The list above has 4 elements which are 'This', 'course', 'is', and 'amazing!'. The original string was separated according to the **white space** between the original string.

3.  Given the variables height and mountain, use .format() to print the following string:
`The height of Mt. Everest is 8848 meters.`

In [ ]:
height = 8848
mountain = "Mt. Everest"

print("The height of {} is {} meters".format(mountain, height))

#first alternative implementation
print("The height of {0} is {1} meters".format(mountain, height))

The height of Mt. Everest is 8848 meters
The height of Mt. Everest is 8848 meters


###Observation:
My observation is that the .format() function replaces the expression {} with the value of the variable inside the parameter of the .format() function. In the alternative implementation, I specified the index of the parameter in the curly brackets which further proves that the .format() function formats the string with the defined variable inside the parameter.



4. Given the following nested list, use indexing to grab the word "this".

```
lst = ['a','b',[4,10,11],['c',[1,66,['this']],2,111],'e',7]
```

In [ ]:
lst = ['a','b',[4,10,11],['c',[1,66,['this']],2,111],'e',7]
print(lst[3][1][2][0])

this


###Observation:
I observed that the given list has a list inside it so to access the word "this" there are multiple indices. We all know that the counting of index in a list starts from 0 so I counted the indices of every element to get the word "this". So I first accessed the element in index 3 and then once inside the element in index 3 which is a list, I then accessed the element in index 1 and then once inside it I accessed the element in index 2 and then lastly since there is only one element in the last list, the index I accessed is 0.

5. Given the following nested dictionary, grab the word "that". This exercise is a little more difficult.
```
d = {'k1':['val1','val2','val3',{'we':['need','to','go',{'deeper':[1,2,3,'that']}]}]}
```

In [ ]:
d = {'k1':['val1','val2','val3',{'we':['need','to','go',{'deeper':[1,2,3,'that']}]}]}

print(d['k1'][3]['we'][3]['deeper'][3])

that


###Observation:
I observed that there are multiple dictionaries inside the original dictionary. To access it, you just need to go inside the values of the dictionary by accessing the key. In this case, the keys are 'k1', 'we' and 'deeper'. I traversed inside the dictionary and accessed the specific indices to grab the word 'that'.

6. Create a function, GetDomain(), that grabs the email website domain from a string in the form: user@domain.com.
So for example, passing "user@domain.com" would return: domain.com

In [ ]:
def GetDomain(x):
  find = "@"
  for i in range(len(x)):
    if(find == x[i]):
      return x[i+1:]

sample = "user@domain.com"
test = GetDomain(sample)
print(test)


domain.com


###Observation:
I observed that in this code, we need to get a specific string after a certain symbol. In this case, we need to get the email website domain which is after the symbol "@". To get the specific website domain everytime after the "@", what I did is traverse the entire string and first find the symbol "@", this can be seen in the conditional statement (find == x[i]). Then after I found the "@" symbol, I sliced the string so that it will return the string after the "@" symbol.

7. Create a basic function, findInternet(), that returns True if the word 'Internet' is contained in the input string. Don't worry about edge cases like punctuation being attached to the word, but account for capitalization. (Hint: Please see https://docs.python.org/2/reference/expressions.html#in)

In [ ]:
def findInternet(x):
  find = "Internet"
  if(find in x):
    return True
  else:
    return False

#First test expected true
sentence = "Internet is amazing and powerful!"
test = findInternet(sentence)
print(test)

#Second test expected false
sentence2 = "Why is the internet so powerful?"
test2 = findInternet(sentence2)
print(test2)

True
False


###Observation:
I observe that in this number, the function "in" finds the specific word in a sentence. If that word is in the sentence, for example "Internet", it will output True and else False.

8. Create a function, countIoT(), that counts the number of times the word "IoT" occurs in a string. Ignore edge cases but take into account capitalization.

In [ ]:
def countIoT(x):
  find = "IoT"
  xnew = x.split()
  count = 0
  for i in xnew:
    if(find == i):
      count+=1
  return count

#expected output is 1
sentence = "Iot IoT iot iOT"
test = countIoT(sentence)
print(test)

#expected output is 5
sentence2 = "IoT IoT IoT IoT IoT"
test2 = countIoT(sentence2)
print(test2)

1
5


###Observation:
In this code, I observed that to find the word "IoT" and count it, we need to split the sentence and compare it to the word "IoT". Everytime the conditional statement (find==i) is satisfied, the variable count increments.

9. Use lambda expressions and the filter() function to filter out words from a list that do not start with the letter 'd'. For example:
```
seq = ['data','salt' ,'dairy','cat', 'dog']
```
should be filtered down to:
```
['data', 'dairy', 'dog']
```

In [ ]:
seq = ['data','salt' ,'dairy','cat', 'dog']
print(list(filter(lambda x: x.startswith('d'),seq)))

['data', 'dairy', 'dog']


###Observation:
I observed that the function filter removes the elements associated with the lambda function. The lambda function is a type of function that can be used for one liner codes. The resulting list only has words that start with letter 'd'.

10. Use lambda expressions and the map() function to convert a list of words to upper case. For example:
```
seq = ['data','salt' ,'dairy','cat', 'dog']
```
should become:
```
['DATA', 'SALT', 'DAIRY', 'CAT', 'DOG']
```

In [ ]:
seq = ['data','salt' ,'dairy','cat', 'dog']
print(list(map(lambda x: x.upper(),seq)))

['DATA', 'SALT', 'DAIRY', 'CAT', 'DOG']


###Observation:
I observed that the map function formats each of the elements of the list with the lambda function. The lambda function here is the x.upper() which makes the string all capitalized.

11. Imagine a smart thermostat that is connected to the door, so that it can detect, in addition to the temperature, when people enter or leave the house. <br><br>Write a function that, if the temperature is lower than 20 degrees Celsius, and there are people in the house (encoded as a boolean value to be passed as a parameter to the function), turns on the heating by returning the string "Heating on". When the temperature reaches 23 degrees or there are no people in the house, it returns the string "Heating off". When none of these conditions are met, the function returns "Do nothing".

In [ ]:
def smartThermostat(temp, people):
  if(temp <= 20 and people == True):
    print("Heating on")
  elif(temp == 23 and people == False):
    print("Heating off")
  else:
    print("Do nothing")

smartThermostat(19, True)
smartThermostat(23, False)
smartThermostat(24, True)


Heating on
Heating off
Do nothing


###Observation:
I observed that this is just a basic use of conditional statements and it can be solved by just using one function.

12. The function zip(list1, list2) returns a list of tuples, where the i-th tuple contains the i-th element from each of the argument lists. Use the zip function to create the following list of tuples: <br>
```
zipped = [('Parking', -1), ('Shops',0), ('Food Court',1), ('Offices',2)]
```

In [ ]:
# Code cell 13

floor_types = ['Parking', 'Shops', 'Food Court', 'Offices']
floor_numbers = range(-1,3)
zipped = list(zip(floor_types, floor_numbers))
#zipped = list(...)

print(zipped)

[('Parking', -1), ('Shops', 0), ('Food Court', 1), ('Offices', 2)]


###Observation:
I observed that the floor types and the floor numbers became the a list inside the tuple.

13. Use the zip function and dict() to create a dictionary, elevator_dict, where the keys are the floor types and the values are the corresponding floor number so that:<br>
```
elevator_dict[-1] = 'Parking'
```

In [ ]:
# Code cell 14
floor_types = ['Parking', 'Shops', 'Food Court', 'Offices']
floors_numbers = range(-1,3)
elevator_dict = dict(zip(floor_numbers,floor_types))
#elevator_dict = dict(...)
print(elevator_dict)

{-1: 'Parking', 0: 'Shops', 1: 'Food Court', 2: 'Offices'}


In [ ]:
# Code cell 15
# Verify elevator_dict[-1]
print(elevator_dict[-1])

Parking


###Observation:
I observed that this code is like the previous one but this is a dictionary and not a list. The floor numbers became the keys and the floor types became the values of the dictionary.



14. Create an Elevator class. The constructor accepts the list of strings floor_types and the list of integers floor_numbers. The class implements the methods 'ask_which_floor' and 'go_to_floor'. The output of this methods should look as follows:
```
floor_types = ['Parking', 'Shops', 'Food Court', 'Offices']
floors_numbers = range(-1,4)
el = Elevator(floor_numbers, floor_types)
el.go_to_floor(1)
Going to Food Court floor!
el.go_to_floor(-2)
There is floor number -2 in this building.
el.ask_which_floor('Offices')
The floor Offices is the number: 2
el.ask_which_floor('Swimming Pool')
There is no Swimming Pool floor in this building.
```

In [ ]:
class Elevator:
  def __init__(self, floor_numbers, floors_types):
    self.FltypesandFlnumbers = dict(zip(floors_numbers, floor_types))
  def go_to_floor(self, floor_number):
    if floor_number in self.FltypesandFlnumbers:
      print("Going to {} floor!".format(self.FltypesandFlnumbers[floor_number]))
    else:
      print("There is no floor number {} in this building".format(floor_number))
  def ask_which_floor(self, floor_type):
    if floor_type in self.FltypesandFlnumbers.values():
        floor_number = {i for i, value in self.FltypesandFlnumbers.items() if value==floor_type}
        print("The floor {} is the number {}".format(floor_type, floor_number))
    else:
      print("There is no floor {} in this building".format(floor_type))

floor_types = ['Parking', 'Shops', 'Food Court', 'Offices']
floors_numbers = range(-1,4)

el = Elevator(floor_numbers, floor_types)
el.go_to_floor(1)
el.go_to_floor(-2)
el.ask_which_floor('Offices')
el.ask_which_floor('Swimming Pool')

Going to Food Court floor!
There is no floor number -2 in this building
The floor Offices is the number {2}
There is no floor Swimming Pool in this building


###Observation:
This code is very difficult for me because I forgot how to create a class and manipulate a dictionary. But with the help of the google, I managed to finish this problem and I successfully showed the value that we need to show. After this HOA, my memory of this language became refreshed and I finally remember all the things I learned in the previous courses.

### Good Job!